### PR TITLE
Add conditional set-me in config

### LINKED ISSUE #1110: [5] Improve config validation using conditional "set-me"s
**Is your feature request related to a problem? Please describe.**
Since the apps config contains a number of optional features, or subfeatures that may require special configuration it would be nice to have some way of validating that the configuration for these are correct.

Currently in our `validate_config` we only validate that a small number of global "set-me"s are set and I think this can be extended.

**Describe the solution you'd like**
I think we should come up with a common standard for conditional "set-me"s that contain say `set-me-if-<yaml.path>=<value>`, and then in our validation function we do a pre-filtering to change these in the merged config to regular "set-me"s if the condition holds, and then run our regular check.

**Describe alternatives you've considered**
This could probably be implemented in some other way, or we could continue with our current way of only commenting out the optional disabled settings hand have no validation.

**Definition of done:**
Validate config can warn if enabled optional features are not configured.

### PR DESCRIPTION
**Pull request by Ajarmar**
> [!warning]
> **This is public repository, ensure not to disclose:**
> - [x] personal data beyond what is necessary for interacting with this pull request
> - [x] business confidential information, such as customer names

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change  
- [ ] kind/dev-change    
- [ ] kind/security      
- [ ] [kind/adr]()       

### What does this PR do / why do we need this PR?

Adds conditional set-me config that will warn the user if the option is not set and the condition is fulfilled.

- Fixes #1110 

#### Additional information to reviewers

- The condition after `set-me-if-` has to be a valid `yq4` condition. I added the parentheses because I felt it made it easier to read, but they are not strictly necessary unless there are multiple conditions.
- I tried adding this for secrets as well but it seems that `set-me`s in the secrets file are not validated at all currently, so it seemed out of scope for this task.
- I checked if it would be possible to add conditions to the json schemas added in [this PR](https://github.com/elastisys/compliantkubernetes-apps/pull/1862) too, and it seems possible but it has to be manually added to the schemas (i.e. it can't just see that an option is conditional in the config and skip schema validation based on that). I didn't add it yet since that PR is not merged yet; if it merges before this one I could add it but otherwise I think it's more appropriate to create a separate task for it. Example for how such a conditional would look in the schema:
  ```yaml
  properties:
    ingressNginx:
      properties:
        controller:
          properties:
            service:
              type: object
              if:
                properties:
                  enabled:
                    const: true
              then:
                properties:
                  type:
                    enum:
                      - ClusterIP
                      - NodePort
                      - LoadBalancer
                      - ExternalName
              properties:
                enabled:
                  type: boolean
                type:
                  type: string
  ```
- When the conditional is fulfilled, the option is changed to `set-me` only in the temporary config file, not in the actual config file. Not sure what the desired outcome would be here but this was more natural with how I implemented it, open for feedback on this.
- I didn't add any migration scripts for this as most `set-me` options should already be dealt with for existing environments, so it wouldn't make much of a difference. Open for feedback on this as well.

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [x] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [x] The metrics are still exposed and present in Grafana after the change
  - [x] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [x] The logs do not show any errors after the change
- Network Policy checks:
  - [x] Any changed pod is covered by Network Policies
  - [x] The change does not cause any dropped packages in the `NetworkPolicy Dashboard`
- Pod Security Policy checks:
  - [x] Any changed pod is covered by Pod Security Admission
  - [x] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [x] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Falco checks:
  - [x] The change does not cause any alerts to be generated by Falco
- Audit checks:
  - [x] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Bug checks:
  - [ ] The bug fix is covered by regression tests

### CODE REVIEW

**Review by robinAwallace — COMMENTED** [review]
I like this change :+1: Could you add unit tests for it? Then I would be happy to approve it :slightly_frowning_face:

**Comment by Ajarmar** [comment]
I pushed unit tests for the conditional set-me functionality. I think some of them perhaps should be split up a bit more, but I wasn't sure about the exact granularity we wanted. Please take a look. (You can try running the tests with `make unit/bin/conditional-set-me/main`.)

I also noticed that some of the old tests are still failing, but I think I will not have time to fix those before I go on vacation.

File: bin/common.bash [hunk 1]
```diff
@@ -344,6 +344,36 @@ validate_version() {
...
 validate_config() {
     log_info "Validating $1 config"
+
+    check_conditionals() {
+        merged_config="${1}"
+        template_config="${2}"
+
+        # Loop all lines in ${template_config} and checks if same option has conditional set-me in ${merged_config}
+        options=$(yq_read_block "${template_config}" "set-me-if-*")
+        maybe_exit="false"
```
Comment (aarnq) [inline bin/common.bash:354]:
```suggestion
```

**Review by aarnq — COMMENTED** [review]
This looks good, though I get three "Do you want to continue" which is a bit too much, can we make it so that it is only printed once after conditionals, validate + schema validate?

**Comment by Zash** [comment]
> I didn't add it yet since [that PR](https://github.com/elastisys/compliantkubernetes-apps/pull/1862) is not merged yet

That PR has been merged.

**Comment by Ajarmar** [comment]
> This looks good, though I get three "Do you want to continue" which is a bit too much, can we make it so that it is only printed once after conditionals, validate + schema validate?

I'm surprised you got three, I only got two (one after validate and one after schema validate). Now there should only be one, unless I missed the thing that made you get three.

Will work on the conditionals for the schema now.

**Review by Zash — APPROVED** [review]
LGTM

File: bin/common.bash [hunk 2]
```diff
@@ -344,24 +344,48 @@ validate_version() {
...
+        template_config="${2}"
+
+        # Loop all lines in ${template_config} and checks if same option has conditional set-me in ${merged_config}
+        options=$(yq_read_block "${template_config}" "set-me-if-*")
+        for opt in ${options}; do
+            opt_value=$(yq4 "${opt}" "${merged_config}")
+            opt_value_no_list=$(echo "${opt_value}" | yq4 "[.] | flatten | .[0]")
+
+            if echo "${opt_value_no_list}" | grep -Pq "^set-me-if-.*$"; then
+                required_condition=$(echo "${opt_value_no_list}" | sed -rn 's/^set-me-if-(.*)/\1/p')
```
[the remark this answers is not in the document]
Comment (Ajarmar) [reply bin/common.bash:359]: Did you try running init? I was hoping that would replace all the pre-existing ones.

[the remark this answers is not in the document]
Comment (Zash) [reply bin/common.bash:359]: Ah, `init` fixes it, never mind then. :)

File: bin/common.bash [hunk 3]
```diff
@@ -344,24 +344,48 @@ validate_version() {
 #       future.
 validate_config() {
     log_info "Validating $1 config"
+
+    check_conditionals() {
+        merged_config="${1}"
+        template_config="${2}"
+
+        # Loop all lines in ${template_config} and checks if same option has conditional set-me in ${merged_config}
+        options=$(yq_read_block "${template_config}" "set-me-if-*")
```
Comment (aarnq) [inline bin/common.bash:353]:
```suggestion
        options="$(yq_read_block "${template_config}" "set-me-if-*")"
```

File: bin/common.bash [hunk 4]
```diff
@@ -344,24 +344,48 @@ validate_version() {
...
+        for opt in ${options}; do
+            opt_value=$(yq4 "${opt}" "${merged_config}")
+            opt_value_no_list=$(echo "${opt_value}" | yq4 "[.] | flatten | .[0]")
```
Comment (aarnq) [inline bin/common.bash:356]:
```suggestion
            opt_value="$(yq4 "${opt}" "${merged_config}")"
            opt_value_no_list="$(yq4 "[.] | flatten | .[0]" <<< "${opt_value}")"
```

File: bin/common.bash [hunk 5]
```diff
@@ -344,24 +344,48 @@ validate_version() {
...
+
+            if echo "${opt_value_no_list}" | grep -Pq "^set-me-if-.*$"; then
```
Comment (aarnq) [inline bin/common.bash:358]:
```suggestion
            if [[ "${opt_value_no_list}" =~ ^set-me-if-.*$ ]]; then
```

File: bin/common.bash (hunk 2 above)
Comment (aarnq) [inline bin/common.bash:359]:
```suggestion
                required_condition="$(sed -rn 's/^set-me-if-(.*)/\1/p' <<< "${opt_value_no_list}")"
```

File: bin/common.bash [hunk 6]
```diff
@@ -344,24 +344,48 @@ validate_version() {
...
+                required_condition=$(echo "${opt_value_no_list}" | sed -rn 's/^set-me-if-(.*)/\1/p')
+                if [[ $(yq4 "${required_condition}" "${merged_config}") == "true" ]]; then
```
Comment (aarnq) [inline bin/common.bash:360]:
```suggestion
                if [[ "$(yq4 "${required_condition}" "${merged_config}")" == "true" ]]; then
```

File: bin/common.bash [hunk 7]
```diff
@@ -344,24 +344,48 @@ validate_version() {
...
+                    # If the option is a list, set the first element in the list
+                    if [[ $(yq4 "${opt} | tag" "${merged_config}") == "!!seq" ]]; then
```
Comment (aarnq) [inline bin/common.bash:362]:
```suggestion
                    if [[ "$(yq4 "${opt} | tag" "${merged_config}")" == "!!seq" ]]; then
```

File: bin/init.bash [hunk 8]
```diff
@@ -86,15 +86,20 @@ generate_sops_config() {
...
 # Usage: replace_set_me <file> <field> <value>
 replace_set_me(){
     if [[ $# -ne 3 ]]; then
         log_error "ERROR: number of args in replace_set_me must be 3. #=[$#]"
         exit 1
     fi
-    if [[ $(yq4 "${2}" "${1}") =~ ^set-me.* ]]; then
+    if [[ $(yq4 "${2}" "${1}") == "set-me" ]]; then
         yq4 --inplace "${2} = ${3}" "${1}"
+    elif [[ $(yq4 "${2}" "${1}") =~ ^set\-me\-if\-.*$ ]]; then
```
Comment (aarnq) [inline bin/init.bash:98]: Do the dashes need to be escaped?

File: config/sc-config.yaml [hunk 9]
```diff
@@ -1182,7 +1182,7 @@ networkPolicies:
   global:
     objectStorageSwift:
       ips:
-        - "set-me-if-enabled"
+        - set-me-if-(.harbor.persistence.type == "swift")
```
Comment (aarnq) [inline config/sc-config.yaml:1185]: Or thanos.

File: bin/init.bash (hunk 8 above)
Comment (Ajarmar) [reply bin/init.bash:98]: Maybe not. In any case, this function is not used anymore after the config template rework, so I'll just remove it.

File: bin/common.bash [hunk 10]
```diff
@@ -376,29 +400,30 @@ validate_config() {
...
         yq_merge "${config_template_path}/common-config.yaml" \
             "${config_template_path}/sc-config.yaml" \
             > "${template_file}"
+        check_conditionals "${config[config_file_sc]}" "${template_file}"
         validate "${config[config_file_sc]}" "${template_file}"
         schema_validate "${config[config_file_sc]}" "${config_template_path}/schemas/config.yaml"
+        check_conditionals "${secrets[secrets_file]}" "${config_template_path}/secrets.yaml"
         validate "${secrets[secrets_file]}" "${config_template_path}/secrets.yaml"
         schema_validate "${secrets[secrets_file]}" "${config_template_path}/schemas/secrets.yaml"
     elif [[ $1 == "wc" ]]; then
```
Comment (robinAwallace) [inline bin/common.bash:429]: Nit: could you extract the common parts for sc/wc and run them after the if-statements?

Something like this?
```suggestion
    if [[ $1 == "sc" ]]; then
        check_config "${config_template_path}/common-config.yaml" \
            "${config_template_path}/sc-config.yaml" \
            "${config_template_path}/secrets.yaml"
        yq_merge "${config_template_path}/common-config.yaml" \
            "${config_template_path}/sc-config.yaml" \
            > "${template_file}"
            
        config_to_validate="${config[config_file_sc]}"
    elif [[ $1 == "wc" ]]; then
    ...
    else
    ...
    fi
    
  	check_conditionals "${config_to_validate}" "${template_file}"
    validate "${config_to_validate}" "${template_file}"
    schema_validate "${config_to_validate}" "${config_template_path}/schemas/config.yaml"
    check_conditionals "${secrets[secrets_file]}" "${config_template_path}/secrets.yaml"
    validate "${secrets[secrets_file]}" "${config_template_path}/secrets.yaml"
    schema_validate "${secrets[secrets_file]}" "${config_template_path}/schemas/secrets.yaml"
```

**Review by robinAwallace — APPROVED** [review]
This looks very nice! Just one nit pick, but it LGTM! :partying_face:

**Pull request merged**